lisa/tests/staging/sched_android.py

```diff
@@ -0,0 +1,290 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2019, Arm Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import os.path
+import abc
+
+from lisa.wlgen.rta import RTA, Periodic
+from lisa.tests.base import ResultBundle, CannotCreateError, RTATestBundle
+from lisa.trace import Trace, FtraceCollector, FtraceConf, requires_events
+from lisa.target import Target
+from lisa.utils import ArtifactPath
+
+class SchedTuneTest(RTATestBundle, abc.ABC):
+    """
+    Abstract class for SchedTune testing
+
+    :param boost: The SchedTune `boost` level to apply the rtapp's cgroup
+    :type boost: int
+
+    :param prefer_idle: The SchedTune `prefer-idle` flag to apply to the rtapp's
+        cgroup
+    :type prefer_idle: bool
+
+    This class provides :meth:`get_cgroup` to create in the SchedTune
+    controller a `lisa_test` cgroup configured with the `boost` and
+    `prefer_idle` attributes.
+    """
+
+    boost = 0
+    prefer_idle = False
+
+    @classmethod
+    def get_cgroup(cls):
+        cg = {
+                'controller': 'schedtune',
+                'name': 'lisa_test',
+                'values': {
+                    'boost': cls.boost,
+                    'prefer_idle': int(cls.prefer_idle)
+                }
+             }
+
+        return cg
+
+class SchedTuneFreqBoost(SchedTuneTest, abc.ABC):
+    """
+    Abstract class for testing SchedTune's frequency biasing feature
+
+    Run a tiny rt-app task pinned to a big CPU in a cgroup with a predefined
+    boost level, and check by looking at `cpu_frequency` events in the trace
+    that the effective frequency was close enough to the expected boosted
+    frequency.
+    """
+
+    ftrace_conf = FtraceConf({
+        "events" : ["sched_switch", "cpu_frequency"]
+    }, __qualname__)
+
+    @classmethod
+    def check_from_target(cls, target):
+        # XXX: better way to do that ?
+        if not target.file_exists('/sys/kernel/debug/tracing/available_events'):
+            raise CannotCreateError("Ftrace not enabled")
+
+        # XXX: better way to do that ?
+        for m in ['cpufreq', 'cgroups', 'sched']:
```

Comment:
That's another one we don't explicitely check (and we probably should). Maybe we could add a class attribute that lists the required devlib modules and automagically check for them? @douglas-raillard-arm ?

Reply:
If we provide a base `RTATestBundle.check_from_target()` it could go in it. That said, it still feels like information duplication: if a required module is not present, `from_target` will fail. If the error it gives is not deemed explicit enough, we should work on that rather than maintaining duplicates that don't introduce any concrete benefit other than presenting the thing in a slightly different way. In real life, we will forget about that list and since all modules are loaded, we will never realize if that list is in sync with the code.

Reply:
Right, if we just want to rely on `from_target` to fail, I think I can remove this function safely and just use the default. We could also remove the schedutil check in `EASBehaviour.check_from_target()` I assume.

Reply:
Hmm not sure I agree with @douglas-raillard-arm. AFAIR that checking step was added to fail early (and explicitely) rather than explode late. So those early checks make sense (at least the cpufreq governors / sched features) IMO.

If that is now deemed superfluous (and I'm not entirely convinced of it), then we should remove the hooks altogether.

Reply:
Then if we follow that logic, we should get rid of `check_from_target`, no? I think it's nice to have some way of checking if running something is going to give me an exception (which is the point of `can_create_from_target()`)

Reply:
Generally speaking I would be for removing `check_from_target` if that means keeping an unmaintained list of modules, with a check that will always succeed in practice since we load them all. AFAIK, "early" means between Target creation and running rtapp. Most if not all TestBundle will access the said features by the time they start running lengthy things. If we really need to do these checks without running rtapp, we can introduce a `check` parameter in `from_target()`:

```
def can_create_from_target(self, *args, **kwargs):
    try:
        self.from_target(*args, **kwargs)
    except CheckPassedException:
        return True
    except Exception:
        return False
    else:
        return True


def check_barrier(raise_):
    if raise_:
        raise CheckPassedException

def from_target(..., check=False):
    # Do most of the setup, using all features that could raise
    ...
    check_barrier(check)
    # Run lengthy things here
    ...
```

That will leave the target set up, but tests need to be able to recover from that anyway, so it would give an easy way to make sure all setup code can work on a non-clean target as it should.

note: there is nothing to be afraid of regarding exceptions, and general python wisdom even says "it's easier to ask for forgiveness than to get permission" == "don't spend time duplicating code and introducing race conditions, just catch exceptions when they occur".

Reply:
I think we should keep `check_from_target()` for things that won't prevent `from_target()` from running but will create unusable artifacts for the test methods.

For example, we can remove the schedutil check from `EASBehaviour.check_from_target()` because `from_target()` will fail anyways if it's not present. But we should keep the `nrg-model` check, because not having it will not prevent `from_target()` from running, but that will break `test_task_placement()`.

So in short, if `from_target()` can run but it's useless because we won't have what we need for the test afterwards, we can just stop early.